src/mod_cache_users.erl

```diff
@@ -0,0 +1,236 @@
+%%% @doc Caches info about non-anonymous users using a queue of ets tables
+-module(mod_cache_users).
+
+-behaviour(gen_mod).
+-behaviour(gen_server).
+
+%% gen_server callbacks
+-export([start_link/2]).
+-export([init/1]).
+-export([handle_call/3]).
+-export([handle_cast/2]).
+-export([handle_info/2]).
+
+%% API
+-export([start/2]).
+-export([stop/1]).
+-export([config_spec/0]).
+-export([supported_features/0]).
+
+-export([does_user_exist/2]).
+
+%% Hooks.
+-export([remove_user/3]).
+-export([remove_domain/3]).
+
+-include("mongoose_config_spec.hrl").
+-include("mongoose.hrl").
+-include("jlib.hrl").
+
+tbl_name() ->
+    mongoose_users.
+
+-spec tbl_name(mongooseim:host_type()) -> atom().
+tbl_name(HostType) ->
+    gen_mod:get_module_proc(HostType, tbl_name()).
+
+%%====================================================================
+%% API
+%%====================================================================
+
+-spec start(mongooseim:host_type(), gen_mod:module_opts()) -> ok.
+start(HostType, Opts) ->
+    start_cache(HostType, Opts),
+    ejabberd_hooks:add(hooks(HostType)),
+    ok.
+
+-spec stop(mongooseim:host_type()) -> ok.
+stop(HostType) ->
+    ejabberd_hooks:delete(hooks(HostType)),
+    stop_cache(HostType),
+    ok.
+
+-spec config_spec() -> mongoose_config_spec:config_section().
+config_spec() ->
+    #section{items = #{
+                       <<"time_to_live">> => #option{type = integer, format = {kv, ttl}},
+                       <<"number_of_segments">> => #option{type = integer}
+                      }}.
+
+-spec supported_features() -> [atom()].
+supported_features() ->
+    [dynamic_domains].
+
+hooks(HostType) ->
+    [
+     {remove_user, HostType, ?MODULE, remove_user, 30},
+     {remove_domain, HostType, ?MODULE, remove_domain, 30}
+    ].
+
+-spec does_user_exist(HostType :: mongooseim:host_type(), JID :: jid:jid()) -> boolean().
+does_user_exist(HostType, #jid{luser = LUser, lserver = LServer} = JID) ->
+    case does_cached_user_exist(HostType, LServer, LUser) of
+        true -> true;
+        false ->
+            case does_stored_user_exist(HostType, JID) of
+                true ->
+                    put_user_into_cache(HostType, LServer, LUser),
+                    true;
+                false -> false
+            end
+    end.
+
+%%====================================================================
+%% Hooks
+%%====================================================================
+
+-spec remove_user(Acc :: mongoose_acc:t(),
+                  LUser :: jid:luser(),
+                  LServer :: jid:lserver() | string()) -> mongoose_acc:t().
+remove_user(Acc, LUser, LServer) ->
```

Comment:
that would not work across the cluster. You have to remove from all nodes (it was broken in the original ejabberd_users version too, btw)

Reply:
Why wouldn't it work? The hook is called on a single node? How would I fix that in all others?

Reply:
The hooks is called on the single node. It's for cleaning database, not for cleaning the local node state.
 Do an rpc call to other nodes. Or send a message.

Reply:
Made a pg2 group.